aptos-move/framework/src/natives/cryptography/algebra/mod.rs

```diff
@@ -169,161 +183,75 @@ macro_rules! safe_borrow_element {
+        _ => None,
     }
 }
 
-/// Macros that implements `deserialize_internal()` using arkworks libraries.
-macro_rules! ark_deserialize_internal {
-    (
-        $gas_params:expr,
-        $context:expr,
-        $structure:expr,
-        $bytes:expr,
-        $format:expr,
-        $ark_typ:ty,
-        $ark_deser_func:ident
-    ) => {{
-        match <$ark_typ>::$ark_deser_func($bytes) {
-            Ok(element) => {
-                let handle = store_element!($context, element);
-                Ok(smallvec![Value::bool(true), Value::u64(handle as u64)])
-            },
-            Err(ark_serialize::SerializationError::InvalidData)
-            | Err(ark_serialize::SerializationError::UnexpectedFlags) => {
-                Ok(smallvec![Value::bool(false), Value::u64(0)])
+#[macro_export]
+macro_rules! abort_unless_arithmetics_enabled_for_structure {
+    ($context:ident, $structure_opt:expr) => {
+        let flag_opt = feature_flag_from_structure($structure_opt);
+        abort_unless_feature_flag_enabled!($context, flag_opt);
+    };
+}
+
+#[macro_export]
+macro_rules! abort_unless_feature_flag_enabled {
+    ($context:ident, $flag_opt:expr) => {
+        match $flag_opt {
+            Some(flag) if $context.get_feature_flags().is_enabled(flag) => {
+                // Continue.
             },
             _ => {
-                abort_invariant_violated();
-                unreachable!()
+                return Err(SafeNativeError::Abort {
+                    abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+                });
             },
         }
-    }};
+    };
 }
 
-fn deserialize_internal(
-    _gas_params: &GasParameters,
-    context: &mut SafeNativeContext,
-    ty_args: Vec<Type>,
-    mut args: VecDeque<Value>,
-) -> SafeNativeResult<SmallVec<[Value; 1]>> {
-    assert_eq!(2, ty_args.len());
-    let structure_opt = structure_from_ty_arg!(context, &ty_args[0]);
-    let format_opt = format_from_ty_arg!(context, &ty_args[1]);
-    abort_unless_serialization_format_enabled!(context, structure_opt, format_opt);
-    let vector_ref = safely_pop_arg!(args, VectorRef);
-    let bytes_ref = vector_ref.as_bytes_ref();
-    let bytes = bytes_ref.as_slice();
-    match (structure_opt, format_opt) {
-        (Some(Structure::BLS12381Fr), Some(SerializationFormat::BLS12381FrLsb)) => {
-            if bytes.len() != 32 {
-                return Ok(smallvec![Value::bool(false), Value::u64(0)]);
-            }
-            ark_deserialize_internal!(
-                gas_params,
-                context,
-                Structure::BLS12381Fr,
-                bytes,
-                SerializationFormat::BLS12381FrLsb,
-                ark_bls12_381::Fr,
-                deserialize_uncompressed
-            )
-        },
-        _ => Err(SafeNativeError::Abort {
-            abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
-        }),
-    }
+fn abort_invariant_violated() -> PartialVMError {
```

Comment:
@vgao1996, we often return this when there's an internal error in our natives, due to us not having maintained the right invariants, or due to a library surprising us with an unexpected return value. Be sure to let us know if this doesn't make sense.